blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/listeners/DownstreamJobListener.java

```diff
@@ -0,0 +1,93 @@
+package io.jenkins.blueocean.listeners;
+
+import hudson.Extension;
+import hudson.model.Run;
+import hudson.model.TaskListener;
+import hudson.model.listeners.RunListener;
+import io.jenkins.blueocean.service.embedded.rest.NodeDownstreamBuildAction;
+import org.jenkinsci.plugins.workflow.cps.nodes.StepStartNode;
+import org.jenkinsci.plugins.workflow.flow.FlowExecution;
+import org.jenkinsci.plugins.workflow.graph.FlowNode;
+import org.jenkinsci.plugins.workflow.job.WorkflowRun;
+import org.jenkinsci.plugins.workflow.support.steps.build.BuildUpstreamNodeAction;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.logging.Logger;
+
+/**
+ * Listens to creation of jobs that are triggered by an upstream job, much like BuildTriggerListener from
+ * pipeline-build-step. Unlike BuildTriggerListener, here we just add an action to the upstream job for later retrieval
+ * rather than injecting information into its build log, in order to be a little more flexible at the cost of a bit more
+ * work later to retrieve the details.
+ */
+@Extension
+public class DownstreamJobListener extends RunListener<Run<?, ?>> {
+
+    private static final Logger LOGGER = Logger.getLogger(DownstreamJobListener.class.getName());
+
+    @Override
+    public void onStarted(Run<?, ?> run, TaskListener listener) {
+
+        for (BuildUpstreamNodeAction action : run.getActions(BuildUpstreamNodeAction.class)) {
+            Run triggerRun = Run.fromExternalizableId(action.getUpstreamRunId());
+            if (triggerRun instanceof WorkflowRun) {
+                FlowExecution execution = ((WorkflowRun) triggerRun).getExecution();
+                FlowNode node;
+
+                if (execution == null) {
+                    LOGGER.warning("Could not retrieve upstream FlowExecution");
+                    continue;
+                }
+
+                try {
+                    node = execution.getNode(action.getUpstreamNodeId());
+                } catch (IOException e) {
+                    LOGGER.warning("Could not retrieve upstream node: " + e);
+                    continue;
+                }
+
+                // Look up the nearest StepStartNode which is where we want to add our action so BO will find it
+                node = findContainingStepStartNode(node);
```

Comment:
`StepStartNode` is ignored and not captured in DAG by blueocean. It only includes stages and parallel branches. Instead what it should do is to walk up ancestors till it finds either parallel branch or a stage. So it could use `PipelineNodeUtil.isStage(node) && PipelineNodeUtils.isParallelBranch(node)` on each ancestor.

Reply:
That didn't fix it either, I'm going to have to collect the actions while building the graph. Which is intimidating, but I've wanted to get a handle on how that works for about a year now, so I'm looking forward to the other side of it :)

Reply:
Oops, should be `PipelineNodeUtil.isStage(node) || PipelineNodeUtils.isParallelBranch(node)`. Anyways, look forward to your proposed changes.

Reply:
>That didn't fix it either
Where ever solution lives, place we want to attach to either stage or parallel node, which ever is immediate parent of triggering step.